Core, Cache: Some love.

FrankFFFF opened the pull request:
:)
Happy April Fool. 
`GMT+8;`

FrankFFFF committed:
Cache: FileEngine need some love.

1. Rename _active to _dirWritable
2. Some tweaks.

markstory commented on src/Cache/Engine/FileEngine.php:
```diff
@@ -312,19 +314,20 @@ protected function _clearDirectory($path, $now, $threshold)
+                    if ($mtime > $threshold) {
+                        continue;
+                    }
+                    $expires = (int)$file->current();
```
Indentation looks wrong.

markstory commented on src/Cache/Engine/FileEngine.php:
```diff
@@ -395,9 +398,7 @@ protected function _setKey($key, $createKey = false)
-            if (!$exists &&
-                !chmod($this->_File->getPathname(), (int)$this->_config['mask'])
-            ) {
+            if (!$exists && !chmod($this->_File->getPathname(), $this->_config['mask'])) {
```
Why remove the int cast? Sometimes people use strings.

markstory commented:
While I know you are trying to improve performance, what kinds of improvements are these changes making to the effected methods?

Fieah commented on src/Core/ObjectRegistry.php:
```diff
@@ -66,25 +66,27 @@
+            if (!empty($config)) {
+                $this->_checkDuplicate($name, $config);
+            }
+
```
I really don't get this, any idea guys?

Fieah commented:
@markstory Hi.
Sorry, I can't express too well in words, i thought the code was clear enough.
English is not my native language. :(

Fieah commented on src/Core/ObjectRegistry.php:
```diff
@@ -110,30 +112,26 @@ protected function _checkDuplicate($name, $config)
-        if (empty($config)) {
-            return;
-        }
+
```
Now we had forced `$config` in `load()` must be an array, so this is unnecessary anymore.

FrankFFFF committed:
Core: ObjectRegistry need some love.

1. load(): Type hinting:  array $config
2. set(): Update `eventManager()->attach` to `eventManager()->on`
3. Some tweaks.

markstory commented:
@Fieah Do the changes you're making improve performance at all? I don't mind merging optimizations, but its good to make sure the changes we're making actually improve performance in measurable ways.

Fieah commented:
@markstory 
Yes, reduce unnecessary check and redundant variable will do.
With my past commits, every little helps :)

markstory commented:
While I agree some of the changes are improvements, adding deeply nested conditionals is arguably 'less clean' by some.

Fieah commented on src/Cache/Engine/FileEngine.php:
```diff
@@ -312,19 +314,19 @@ protected function _clearDirectory($path, $now, $threshold)
+                    if ($expires > $now) {
+                        continue;
+                    }
                 }
```
Hi @markstory 
You mean this changes?

markstory replied:
Yeah, you're at 3 layers of conditionals.

jadb commented:
Please, lets keep the code readable. IMPO, performance, while being a very important thing, if not substantially improved, is not more important than code simplicity/readability.

Fieah commented:
I actually don't get which changes made the code become unreadable,
can someone point it out please?

If readability really that important to this project, than just close this PR.
But still, IMPersonalO, this is just fine.

markstory commented:
@Fieah Readability is always very important for a long running project like this. Unless optimizations yield measurable improvements, I would prefer to keep code more readable. That's why I asked if you had benchmarked the modified methods before/after. If there are measurable performance improvements than I'm willing to sacrifice readability a bit.

FrankFFFF committed:
Cache: FileEngine: Remove checking $file->isFile()

SplFileObject::construc() does the trick.

Fieah commented:
That's unnecessary to do a benchmark, it won't be a big improvement.
I have updated the changes, kindly take a look, thank you :)

Fieah commented:
... The other methods of FileEngine is not using Exceptions for SplFileObject.
Quick question, shorter code or faster response? Sometimes we can't have both.

jadb commented on src/Cache/Engine/FileEngine.php:
```diff
@@ -413,8 +411,16 @@ protected function _setKey($key, $createKey = false)
      * @return bool
      */
-    protected function _active()
+    protected function _dirWritable()
```
Why change the method name?

Fieah replied:
Hmm.. `active` sounds more like from a config to me.

rchavik replied:
Please do not rename methods in a bugfix release.

Fieah replied:
This is a protected method, and it is only use for once.

rchavik replied:
There's always a possibility that something extends this class and uses that method.

jadb replied:
@rchavik is totally right. Same goes about method signature at such low levels like ObjectRegistry.

ceeram replied:
protected methods can be extended outside the framework and userland code will break if methods are renamed in minor or bugfix releases. BC can only be broken in major versions.

jadb commented on src/Core/ObjectRegistry.php:
```diff
@@ -66,25 +66,27 @@
      * @return mixed
      */
-    public function load($objectName, $config = [])
+    public function load($objectName, array $config = [])
```
Not so keen on forcing the array type. Some classes will want to extend the `ObjectRegistry` and offer a quick way of for example passing the class name:

```php
public function load($objectName, $config = []) {
    if (is_string($config)) {
        $config = ['className' => $config];
    }
    ...
}
```

Fieah replied:
Actually this is bugging me too, please take a look at `_checkDuplicate` method.
`
        unset($config['enabled'], $existingConfig['enabled']);

        $fail = false;
        foreach ($config as $key => $value) {
`

Fieah replied:
And also we use array as typehint in description and default value, so...

dereuromark replied:
Those are not typehints, though. So this could also break existing apps. We need to stick to the non-typehint for 3.0 and we can rethink it for 3.1.

jadb replied:
What about it? Please be clear just so we don't make assumptions. Other are reading and will reading those comments in the future.

Fieah replied:
Hmm.. maybe i was wrong or the cook book needs a correction.
http://book.cakephp.org/3.0/en/contributing/cakephp-coding-conventions.html#typehinting

Fieah replied:
Read the `_checkDuplicate`, and try pass in a string.
**Boogeyman is in the corner** 
Don't just talk about the typehints.

jadb commented on src/Core/ObjectRegistry.php:
```diff
@@ -281,7 +279,7 @@ public function set($objectName, $object)
         $this->unload($objectName);
         if (isset($this->_eventManager)) {
-            $this->eventManager()->attach($object);
+            $this->eventManager()->on($object);
```
:+1:

jadb commented:
@Fieah, I like how now the code in general is even more readable than it initially was (performance increase or not), good job!

jippi commented:
I don't "get" half these changes - also the ticket title does not make sense to the actual diff, there is no eval()

jippi commented:
Also, would like some actual benchmarks on how this improve performance - decreasing readability should only be done if the performance gain is fairly big imo

dereuromark commented:
@Fieah To avoid this kind of pro/contra it probably would have helped if you divided those changes into separate small PRs, each outlining exactly one type of change (with the appropriate commit message).
I like the moving of code further down where a "return early" would render this call unnecessary in the first place. I also like some other changes. But as @jippi pointed out, some others are just noise in the diff or decrease readability. Those make merging the good changes kind of difficult then.
But still: You are on the right track.

Fieah commented:
The title was from yesterday, Arpril 1st.

I HATE to sounds like a d**k, but if any changes from this PR could make users confused, 
then maybe he/she is just not READY.

Please, someone read the outdated diff and the notes above,
and please kindly add a note to the line you think it was decrease the readability. 
Thank you.

I'm ran out of words, and now I'm hating myself for not taking my english class seriously.

dereuromark commented on src/Core/ObjectRegistry.php:
```diff
@@ -110,30 +112,26 @@ protected function _checkDuplicate($name, $config)
+            if (!isset($existingConfig[$key])) {
                 $fail = true;
                 break;
+            } else {
```
the else is not necessary when you return/break early.

Fieah replied:
Noted, thank you, @dereuromark :)

dakota commented:
I'm a big fan of Object Calisthenics, especially the "one level of indentation rule" ( https://github.com/object-calisthenics/phpcs-calisthenics-rules#rule-1-only-one-level-of-indentation). The basic idea being that you should never have more than a single level of indentation. Here's a fantastic talk given at a past CakeFest about it: https://www.youtube.com/watch?v=GtB5DAfOWMQ

ceeram commented:
Even though love for the code is always appreciated, i agree with earlier comments. Without some benchmarks showing significant perfomance improvements, we should avoid additional nesting and readability. I guess you can love code in different ways ;)
Your efforts on improving cakephp codebase is highly appreciated, even if we do not disagree on all changes. Don't let it stop you showing your love :)
big :+1: on your efforts, but im :-1: on making code less readable.

ceeram commented on src/Cache/Engine/FileEngine.php:
```diff
@@ -314,24 +316,22 @@ protected function _clearDirectory($path, $now, $threshold)
                     continue;
                 }
             }
-            if ($file->isFile()) {
```
Are you sure, this check isnt needed anymore?

Fieah replied:
if `SplFileObject($path . $entry, 'r')` went wrong, it `continue`
No matter `$path . $entry` is not exists or unreadable.

Fieah commented:
PR closed.
I'm so tired to "..." whatever this is with 7 different persons.

But still, thanks a lot for your time.

Fieah commented:
Again.
If any changes from **this PR** could make users confused, 
then maybe this guy is just **not READY, and should just go play with HTML.**
I understand about the readability is important, but not in this case.

BTW,
funny story, i don't even had Cake installed before or now.
All of my PR was only been read and edited directly then test with CI in here.
This explain a lot for #6190 , right? 

Anyway, PR is closed, but IMO, some changes that i had noted above is necessary, 
that's the real thing will confused users.

That's all.
Have a nice day.

AD7six commented:
>  some changes that i had noted above is necessary

The lesson there is to _not_ mix required/functional changes with nice-to-have/code-style/cosmetic changes.

josegonzalez commented:
Note: no one is saying you should stop contributing or that you are not intelligent and therefore should "play with something lesser" than cake (that assumes HTML is easy, its not). We just disagreed about some of your stylistic changes which go against how we generally write code, that's all.

Regardless, thanks for the contributions, @markstory has opened a pr with a few of your changes included.
